### DISCUSSION TITLE
Powerwall system (not plus) with tesla inverter and enphase.  ANSWERED

### DISCUSSION THREAD

**Discussion by anderep** [jasonacox/Powerwall-Dashboard#254] [General]
My solar, tesla and enphase, is monitored by a neurio on individual CTs. Is there a way to access and display the information individually?  I would like a separate out tesla inverter and enphase graphs, along with total.

**Comment by jasonacox** [comment]
Do you see those data points in the vitals response from the gateway (http://localhost:8675/vitals)?  My Neurio shows up:

```json
{
    "NEURIO--VAH4600AB2155": {
        "NEURIO_CT0_InstRealPower": -4.840000152587891,
        "NEURIO_CT0_Location": "solarRGM",
        "componentParentDin": "STSTSM--xxxxxxx-00-E--xxxxxxx",
        "firmwareVersion": "1.7.2-Tesla",
        "lastCommunicationTime": 1682303437,
        "manufacturer": "NEURIO",
        "meterAttributes": {
            "meterLocation": [
                5
            ]
        },
        "serialNumber": "xxx"
    },
```

**Reply by anderep** [reply]
Yes CT0 and CT3 on neurio labeled solar on top of vitals.  Did wget, and it wrote the file out to a file ‘vitals’.


`eric@opensprinkler:~ $ cat vitals 
```groovy
{
    "NEURIO--VAH4635AB1496": {
        "NEURIO_CT0_InstRealPower": 2373.679931640625,
        "NEURIO_CT0_Location": "solar",
        "NEURIO_CT3_InstRealPower": 11851.8203125,
        "NEURIO_CT3_Location": "solar",
        "componentParentDin": "STSTSM--1232100-00-H--GF22227000014F",
        "firmwareVersion": "1.7.2-Tesla",
        "lastCommunicationTime": 1682369955,
        "manufacturer": "NEURIO",
        "meterAttributes": {
            "meterLocation": [
                4,
                4
            ]
        },
        "serialNumber": "VAH4635AB1496"
    },
    "NEURIO--VAH5037AB3569": {
        "NEURIO_CT0_InstRealPower": 2.7300000190734863,
        "NEURIO_CT0_Location": "site",
        "NEURIO_CT1_InstRealPower": 2.309999942779541,
        "NEURIO_CT1_Location": "site",
        "componentParentDin": "STSTSM--1232100-00-H--GF22227000014F",
        "firmwareVersion": "1.7.2-Tesla",
        "lastCommunicationTime": 1682369955,
        "manufacturer": "NEURIO",
        "meterAttributes": {
            "meterLocation": [
                1
            ]
        },
        "serialNumber": "VAH5037AB3569"
    },

**Reply by jasonacox** [reply]
Nice!  There are two devices with 4 CTs.  Telegraf could be set up to pull these in (see `telegraf.local.conf`). Vitals may be an overkill so some telegraf processing to break the different VAH* and NEURIO_CTx values out would be ideal.  We would also need to add a CQ to `influx/influxdb.sql` for long term resampling and retention.

```
"NEURIO--VAH4635AB1496": {
        "NEURIO_CT0_InstRealPower": 2373.679931640625,
        "NEURIO_CT0_Location": "solar",
        "NEURIO_CT3_InstRealPower": 11851.8203125,
        "NEURIO_CT3_Location": "solar",
        "componentParentDin": "STSTSM--1232100-00-H--GF22227000014F",
        "firmwareVersion": "1.7.2-Tesla",
        "lastCommunicationTime": 1682369955,
        "manufacturer": "NEURIO",
        "meterAttributes": {
            "meterLocation": [
                4,
                4
            ]
        },
        "serialNumber": "VAH4635AB1496"
    },
    "NEURIO--VAH5037AB3569": {
        "NEURIO_CT0_InstRealPower": 2.7300000190734863,
        "NEURIO_CT0_Location": "site",
        "NEURIO_CT1_InstRealPower": 2.309999942779541,
        "NEURIO_CT1_Location": "site",
        "componentParentDin": "STSTSM--1232100-00-H--GF22227000014F",
        "firmwareVersion": "1.7.2-Tesla",
        "lastCommunicationTime": 1682369955,
        "manufacturer": "NEURIO",
        "meterAttributes": {
            "meterLocation": [
                1
            ]
        },
        "serialNumber": "VAH5037AB3569"
    },
```

**Reply by anderep** [reply]
first is the solar, enphase an tesla inverters; second is the non backup panel. 

1) where do I begin to break the different VAH* and NEURIO_CTx values out?
2) how do I add a CQ to influx/influxdb.sql for long term resampling and retention?

edit: I did figure out curl grep 

eric@opensprinkler:~ $ curl -s http://localhost:8675/vitals | grep InstRealPower
        "NEURIO_CT0_InstRealPower": 2757.199951171875,
        "NEURIO_CT3_InstRealPower": 16117.3798828125,
        "NEURIO_CT0_InstRealPower": 3813.43994140625,
        "NEURIO_CT1_InstRealPower": 3817.3798828125,
        "METER_X_CTA_InstRealPower": -3974.0,
        "METER_X_CTB_InstRealPower": -3619.0,
        "METER_X_CTC_InstRealPower": 0.0,
        "METER_Y_CTA_InstRealPower": 0.0,
        "METER_Y_CTB_InstRealPower": 0.0,
        "METER_Y_CTC_InstRealPower": 0.0,

**Reply by jasonacox** [reply]
Okay, here is an attempt. I only have one meter so I could be missing something here.  You have two Neurio's with 4 total CT meters so I added what I think you need. 

## Import Neurio Data

>"NEURIO_CT0_InstRealPower": 2757.199951171875,
"NEURIO_CT3_InstRealPower": 16117.3798828125,
"NEURIO_CT0_InstRealPower": 3813.43994140625,
"NEURIO_CT1_InstRealPower": 3817.3798828125,

Add this to `telegraph.local`:

```
# Neurio Data for VAH4635AB1496
[[inputs.http]]
      urls = [
              "http://pypowerwall:8675/vitals"
      ]
      name_override = "neurio0"
      method = "GET"
      insecure_skip_verify = true
      timeout = "4s"
      data_format = "json"
      json_query = "NEURIO--VAH4635AB1496*"

# Neurio Data for VAH5037AB3569
[[inputs.http]]
      urls = [
              "http://pypowerwall:8675/vitals"
      ]
      name_override = "neurio1"
      method = "GET"
      insecure_skip_verify = true
      timeout = "4s"
      data_format = "json"
      json_query = "NEURIO--VAH5037AB3569*"

[[processors.date]]
        tag_key = "month"
        date_format = "Jan"
        timezone = "America/Los_Angeles"

[[processors.date]]
        tag_key = "year"
        date_format = "2006"
        timezone = "America/Los_Angeles"
```

Restart telegraph:  
```bash
docker stop telegraf; docker rm telegraf; ./compose-dash.sh up -d
```

If you run a telegraf test (`docker exec -it telegraf telegraf --config /etc/telegraf/telegraf.d/local.conf  --test`) you should get something like this (but with 2 lines):
```
neurio0,host=telegraf,month=Apr,url=http://pypowerwall:8675/vitals,year=2023 NEURIO_CT0_InstRealPower=3614.739990234375,lastCommunicationTime=1682895075,meterAttributes_meterLocation_0=5 1682895075000000000
```

## Create InfluxDB CQs

To create long term (downsampling) storage of this data in InfluxDB, we need to create a CQ. Create the file `influxdb/neurio.sql` to transfer the NEURIO data into the neurio bucket with fields `neurio_1` ... `neurio_4` representing the 4 meters you have.

```sql
# USE powerwall
CREATE DATABASE powerwall
CREATE RETENTION POLICY neurio ON powerwall duration 0s replication 1
CREATE CONTINUOUS QUERY cq_neurio0 ON powerwall BEGIN SELECT mean("NEURIO_CT0_InstRealPower") AS neurio_1, mean("NEURIO_CT3_InstRealPower") AS neurio_2 INTO powerwall.neurio.:MEASUREMENT FROM (SELECT NEURIO_CT0_InstRealPower FROM "raw"."neurio0" ) GROUP BY time(1m), month, year fill(linear) END
CREATE CONTINUOUS QUERY cq_neurio1 ON powerwall BEGIN SELECT mean("NEURIO_CT0_InstRealPower") AS neurio_3, mean("NEURIO_CT1_InstRealPower") AS neurio_4 INTO powerwall.neurio.:MEASUREMENT FROM (SELECT NEURIO_CT0_InstRealPower FROM "raw"."neurio1" ) GROUP BY time(1m), month, year fill(linear) END
```

Have InfluxDB add the CQs:

```bash
docker exec --tty influxdb sh -c "influx -import -path=/var/lib/influxdb/neurio.sql"
```

You can then graph the Neurio data using  something like `SELECT mean("neurio_1") FROM "neurio"."neurio0" WHERE $timeFilter GROUP BY time($__interval) fill(null)`:

<img width="572" alt="image" src="https://user-images.githubusercontent.com/836718/235379932-63cb8475-f754-4d99-a796-f8ef4ae762fc.png">

Yes, I'm sure it can be optimized and cleaned up a bit. 😉  Let us know if you get something to work.

## More Meter Data

> "METER_X_CTA_InstRealPower": -3974.0,
"METER_X_CTB_InstRealPower": -3619.0,
"METER_X_CTC_InstRealPower": 0.0,
"METER_Y_CTA_InstRealPower": 0.0,
"METER_Y_CTB_InstRealPower": 0.0,
"METER_Y_CTC_InstRealPower": 0.0,

You can use a similar method to pull these in as well.

**Comment by anderep** [comment]
Edited telegraph.local, and restarted docker.  Ran test:


```
eric@opensprinkler:~/Powerwall-Dashboard $ docker stop telegraf; docker rm telegraf; ./compose-dash.sh up -d
telegraf
telegraf
Running Docker Compose...
pypowerwall is up-to-date
influxdb is up-to-date
weather411 is up-to-date
grafana is up-to-date
Creating telegraf ... done
eric@opensprinkler:~/Powerwall-Dashboard $ docker exec -it telegraf telegraf --config /etc/telegraf/telegraf.d/local.conf  --test
2023-05-02T03:27:49Z I! Loading config file: /etc/telegraf/telegraf.d/local.conf
2023-05-02T03:27:49Z I! Starting Telegraf 1.26.1
2023-05-02T03:27:49Z I! Available plugins: 235 inputs, 9 aggregators, 27 processors, 22 parsers, 57 outputs, 2 secret-stores
2023-05-02T03:27:49Z I! Loaded inputs: http (2x)
2023-05-02T03:27:49Z I! Loaded aggregators: 
2023-05-02T03:27:49Z I! Loaded processors: date (2x)
2023-05-02T03:27:49Z I! Loaded secretstores: 
2023-05-02T03:27:49Z W! Outputs are not used in testing mode!
2023-05-02T03:27:49Z I! Tags enabled: host=telegraf
> neurio1,host=telegraf,month=May,url=http://pypowerwall:8675/vitals,year=2023 NEURIO_CT0_InstRealPower=2.799999952316284,NEURIO_CT1_InstRealPower=2.2100000381469727,lastCommunicationTime=1682998065,meterAttributes_meterLocation_0=1 1682998070000000000
> neurio0,host=telegraf,month=May,url=http://pypowerwall:8675/vitals,year=2023 NEURIO_CT0_InstRealPower=-4.440000057220459,NEURIO_CT3_InstRealPower=-1.7799999713897705,lastCommunicationTime=1682998064,meterAttributes_meterLocation_0=4,meterAttributes_meterLocation_1=4 1682998070000000000

```


Created influxdb/neurio.sql and add the QC’s

then got stuck at graphing at grafana…

**Reply by jasonacox** [reply]
Create a new panel and use the drop downs to fill in the FROM, SELECT, GROUP BY, and ALIAS details like this: 

<img width="823" alt="image" src="https://user-images.githubusercontent.com/836718/235578115-8862798c-b0bf-49c8-8223-8cdb6fe5da45.png">

**Comment by anderep** [comment]
i'm only getting two of the four CT
![image](https://user-images.githubusercontent.com/12971922/235605467-621ed76c-0302-420f-bf7f-025fb41e6b6c.png)

**Comment by anderep** [comment]
neurio0/neurio_1 is the tesla inverter, now to get CT3
![image](https://user-images.githubusercontent.com/12971922/235682328-5d7f7938-20f8-4607-9a49-8cff7494f823.png)

**Reply by jasonacox** [reply]
Nice!

Are you not seeing the neurio_2 in the dropdown for `neurio0`?

The CQs you can see the mappings (make sure these are right):

```
CREATE CONTINUOUS QUERY cq_neurio0 ON powerwall BEGIN 
    SELECT 
    mean("NEURIO_CT0_InstRealPower") AS neurio_1,
    mean("NEURIO_CT3_InstRealPower") AS neurio_2 
INTO powerwall.neurio.:MEASUREMENT 
    FROM (SELECT NEURIO_CT0_InstRealPower FROM "raw"."neurio0" ) GROUP BY time(1m), month, year fill(linear) END

CREATE CONTINUOUS QUERY cq_neurio1 ON powerwall BEGIN 
    SELECT 
    mean("NEURIO_CT0_InstRealPower") AS neurio_3, 
    mean("NEURIO_CT1_InstRealPower") AS neurio_4 
INTO powerwall.neurio.:MEASUREMENT 
    FROM (SELECT NEURIO_CT0_InstRealPower FROM "raw"."neurio1" ) GROUP BY time(1m), month, year fill(linear) END

```

**Reply by anderep** [reply]
That is what I have…. Only seem to be pulling one ct from each neurio.
Do I need two more “from” lines?

**Reply by jasonacox** [reply]
Yes!  My bad.  I only set these up to work with my config. I think you can update to:

```sql
# first one
FROM (SELECT NEURIO_CT0_InstRealPower, NEURIO_CT3_InstRealPower FROM "raw"."neurio0" ) GROUP BY time(1m), month, year fill(linear) END

# second one
FROM (SELECT NEURIO_CT0_InstRealPower, NEURIO_CT1_InstRealPower FROM "raw"."neurio1" ) GROUP BY time(1m), month, year fill(linear) 
```

**Reply by anderep** [reply]
Not getting second CT in drop menu.

**Reply by anderep** [reply]
```
$ cat /home/eric/Powerwall-Dashboard/influxdb/neurio.sql 
# USE powerwall
CREATE DATABASE powerwall 
CREATE RETENTION POLICY neurio ON powerwall duration 0s replication 1 
CREATE CONTINUOUS QUERY cq_neurio0 ON powerwall BEGIN SELECT mean("NEURIO_CT0_InstRealPower") AS neurio_1, mean("NEURIO_CT3_InstRealPower") AS neurio_2 INTO powerwall.neurio.:MEASUREMENT FROM (SELECT NEURIO_CT0_InstRealPower, NEURIO_CT3_InstRealPower FROM "raw"."neurio0" ) GROUP BY time(1m), month, year fill(linear) END 
CREATE CONTINUOUS QUERY cq_neurio1 ON powerwall BEGIN SELECT mean("NEURIO_CT0_InstRealPower") AS neurio_3, mean("NEURIO_CT1_InstRealPower") AS neurio_4 INTO powerwall.neurio.:MEASUREMENT FROM (SELECT NEURIO_CT0_InstRealPower, NEURIO_CT1_InstRealPower FROM "raw"."neurio1" ) GROUP BY time(1m), month, year fill(linear) END

```

**Reply by anderep** [reply]
Maybe it isn’t reloading since it exists?
```
$ docker exec --tty influxdb sh -c "influx -import -path=/var/lib/influxdb/neurio.sql"
2023/05/05 02:38:30 error: continuous query already exists
2023/05/05 02:38:30 error: continuous query already exists

```

**Reply by jasonacox** [reply]
Yes, good point! You can drop the old ones and rerun:

```sql
DROP CONTINUOUS QUERY cq_neurio0 ON powerwall
DROP CONTINUOUS QUERY cq_neurio1 ON powerwall
```

**Comment by anderep** [comment]
that was it, thank man!  put the drop commands in a file called drop.sql and ran the docker import command above.  then re-imported the neurio continouse query, and it took.
![image](https://user-images.githubusercontent.com/12971922/236386259-25ae7328-7f80-4ae2-844a-eb1e9bbc132b.png)
my tesla array is 12x 400watt panels on north side of house; the enphase is 63x 320watt panels on east, south, and west side of house.

**Reply by jasonacox** [reply]
Awesome!

**Comment by anderep** [comment]
added up L1 and L2, added current power, and integrated power to get kWh 
![image](https://user-images.githubusercontent.com/12971922/236564263-518295a9-611b-4125-9b3a-38cca6ffffb0.png)
non backup load was charging the MY.
Big load at night is water the lawn...
![image](https://user-images.githubusercontent.com/12971922/236565044-f1c44d41-22c4-4298-8c1a-02665ae9959e.png)

**Reply by jasonacox** [reply]
Wow!  Nice system @anderep

**Comment by anderep** [comment]
i want to graph main panel load (backed up panel).  so house load minus non-backed up panel...  any ideas?

this didn't work, perhaps a global variable...  or maybe backed up panel load is already available...
SELECT mean("home") - mean("neurio_3") - mean("neurio_4") FROM "autogen"."http" "neurio"."neurio1" WHERE $timeFilter GROUP BY time($__interval) fill(null)

**Reply by jasonacox** [reply]
What does it look like when you graph them separate?

**Comment by anderep** [comment]
Non Backup Load:
SELECT mean("neurio_3") + mean("neurio_4") FROM "neurio"."neurio1" WHERE $timeFilter GROUP BY time($__interval) fill(null)

![image](https://user-images.githubusercontent.com/12971922/236625431-ec85ce4c-a365-402a-9e3d-13a322e66269.png)

and i stole home for your dashboard:
SELECT mean("home") FROM "autogen"."http" WHERE $timeFilter GROUP BY time($__interval) fill(null)

![image](https://user-images.githubusercontent.com/12971922/236625641-ead77f69-7050-4914-8f96-fe45f7155bda.png)

**Reply by jasonacox** [reply]
> this didn't work, perhaps a global variable... or maybe backed up panel load is already available...
SELECT mean("home") - mean("neurio_3") - mean("neurio_4") FROM "autogen"."http" "neurio"."neurio1" WHERE $timeFilter GROUP BY time($__interval) fill(null)

Your graphs looks great!  Assuming the NEURIO is "A" and HOME is "B", you could create an `Expression` that is A-B:

<img width="701" alt="image" src="https://user-images.githubusercontent.com/836718/236633513-b25b0025-8134-4b5b-a63b-80cf30aaf956.png">

**Comment by anderep** [comment]
can i pull form other panels, "home usage" is in the "energy usage" panel below?  
`$A - "Energy Usage"$A` 
or
`$A - $"Home Usage" `
????
but it works within one panel.  can i hide a query, so it doesn't graph?  edit:  looks like a can disable a query.  lets see if that works?

![image](https://user-images.githubusercontent.com/12971922/236642490-a24669e0-5c68-4468-98d4-4d2053c22f21.png)
